python-sqlite-sqlalchemy/project/examples/example_2/main.py

```diff
@@ -1,77 +1,185 @@
+        session: database session to use
+        ascending: direction to sort the results
 
-def main():
-    print("starting")
+    Returns:
+        List: list of publisher sorted by number of authors published
+    """
+    if not isinstance(ascending, bool):
+        raise ValueError(f"Sorting value invalid: {ascending}")
+
+    direction = asc if ascending else desc
+
+    return (
+        session.query(
+            Publisher.name,
+            func.count(Author.first_name).label("total_authors"),
+        )
+        .join(Publisher.authors)
+        .group_by(Publisher.name)
+        .order_by(direction("total_authors"))
+    )
+
+
+def get_authors(session):
+    """Get a list of author objects"""
+    return session.query(Author).order_by(Author.last_name).all()
+
+
+def add_new_book(session, author_name, book_title, publisher_name):
+    """Adds a new book to the system"""
+
+    # Get the author if exists
+    first_name, last_name = author_name.split(" ")
+    author = (
+        session.query(Author)
+        .filter(
+            and_(
+                Author.first_name == first_name, Author.last_name == last_name
+            )
+        )
+        .one_or_none()
+    )
+
+    # Get the book if exists
+    book = session.query(Book).filter(Book.title == book_title).one_or_none()
+
+    # Get the publisher if exists
+    publisher = (
+        session.query(Publisher)
+        .filter(Publisher.name == publisher_name)
+        .one_or_none()
+    )
+    # Does book not exist, and the author or publisher also not existexist?
+    if book is not None and not (author is None or publisher is None):
+        raise Exception(
+            "New item exists", author_name, book_title, publisher_name
+        )
+    # Create the book
+    book = Book(title=book_title)
+
+    # Add the book to the author's books collection if didn't exist
+    author.books.append(book)
+
+    # Add the book to the publisher's collection if didn't exist
+    publisher.books.append(book)
+
+    # Commit to the database
+    session.commit()
+
+
+def output_author_hierarchy(authors):
+    """
+    Outputs the author/book/publisher information in
+    a hierarchical manner
+
+    :param authors:         the collection of root author objects
+    :return:                None
+    """
+    authors_tree = Tree()
+    authors_tree.create_node("Authors", "authors")
+    for author in authors:
+        authors_tree.create_node(
+            f"{author.first_name} {author.last_name}",
+            f"{author.first_name} {author.last_name}",
+            parent="authors",
+        )
+        for book in author.books:
+            authors_tree.create_node(
+                f"{book.title}",
+                f"{book.title}",
+                parent=f"{author.first_name} {author.last_name}",
+            )
+            for publisher in book.publishers:
+                authors_tree.create_node(
+                    f"{publisher.name}", uuid4(), parent=f"{book.title}"
```

Comment:
The default is to create a `uuid4()` if the second argument is not given. To be consistent with Example 1 and avoid the extra import, I suggest that we remove  the explicit call to `uuid4()` here:

```suggestion
                    f"{publisher.name}", parent=f"{book.title}"
```